app_dart/lib/src/request_handlers/get_timeseries_history.dart

```diff
@@ -0,0 +1,117 @@
+import '../service/datastore.dart';
+
+@immutable
+class GetTimeSeriesHistory extends RegularRequestHandler<GetTimeSeriesHistoryResponse> {
+  const GetTimeSeriesHistory(
+    Config config, {
+    @visibleForTesting DatastoreServiceProvider datastoreProvider,
+  })  : datastoreProvider =
+            datastoreProvider ?? DatastoreService.defaultProvider,
+        super(config: config);
+
+  final DatastoreServiceProvider datastoreProvider;
+  static const String timeSeriesKeyParam = 'TimeSeriesKey';
+  static const String startFromParam = 'StartFrom';
+
+  @override
+  Future<GetTimeSeriesHistoryResponse> post() async {
+    checkRequiredParameters(<String>[timeSeriesKeyParam]);
+    const int maxRecords = 1500;
+    final DatastoreService datastore = datastoreProvider();
+    final KeyHelper keyHelper = KeyHelper(applicationContext: AppEngineContext(false,'','','','','',Uri()));
+    final Set<Commit> commits =
+        await datastore.queryRecentCommits(limit: maxRecords).toSet();
+
+    Key timeSeriesKey;
+    try {
+      timeSeriesKey = keyHelper.decode(requestData[timeSeriesKeyParam]);
+    } catch (error) {
+      throw BadRequestException(
+          'Bad timeSeries key: ${requestData[timeSeriesKeyParam]}');
+    }
+
+    final TimeSeries timeSeries =
+        await config.db.lookupValue<TimeSeries>(timeSeriesKey, orElse: () {
+      throw BadRequestException('No such timeseries: ${timeSeriesKey.id}');
+    });
+
+    List<TimeSeriesValue> timeSeriesValues = await datastore
+        .queryRecentTimeSeriesValues(timeSeries,
+            startFrom: startFromParam, limit: maxRecords)
+        .toList();
+    timeSeriesValues = insertMissingTimeseriesValues(timeSeriesValues, commits);
+
+    return GetTimeSeriesHistoryResponse(timeSeries, timeSeriesValues);
+  }
+}
+
+@immutable
+class GetTimeSeriesHistoryResponse extends JsonBody {
+  const GetTimeSeriesHistoryResponse(this.timeSeries, this.timeSeriesValues)
+      : assert(timeSeries != null);
+
+  final TimeSeries timeSeries;
+  final List<TimeSeriesValue> timeSeriesValues;
+
+  @override
+  Map<String, dynamic> toJson() {
+    return <String, dynamic>{
+      'BenchmarkData': BenchmarkData(
+          timeSeriesEntity: TimeseriesEntity(timeSeries: timeSeries),
+          values: timeSeriesValues),
+      //LastPosition to do @Keyong
+      'LastPosition': null,
+    };
+  }
+}
+
+List<TimeSeriesValue> insertMissingTimeseriesValues(
+    List<TimeSeriesValue> timeSerialsValues, Set<Commit> commits) {
+  final List<TimeSeriesValue> values = <TimeSeriesValue>[];
+  final Map<String, TimeSeriesValue> valuesByCommit =
+      <String, TimeSeriesValue>{};
+  for (TimeSeriesValue value in timeSerialsValues) {
+    valuesByCommit[value.revision] = value;
+  }
+
+  for (Commit commit in commits) {
+    TimeSeriesValue value;
+
+    if (valuesByCommit.containsKey(commit.sha)) {
+      value = valuesByCommit[commit.sha];
+      if (value.value < 0) {
+        value.value = 0;
+      }
```

Comment:
Why negative value should be regulated to zero?

Reply:
Same from go.

Reply:
Please leave a comment saying this behavior is copied from the Go code.

Reply:
Now that everything on dart side is exactly a direct mirror from GO side. Actually there are a lot of other similar places: such as the maxNum for commits/taskvalues of get-benchmark API/get-timeseries-history API. Another thing to mention is that our existing cocoon backend currently has quite a few places to improve, and our next step after the migration finishes is to focus on improvement. Since we assume everything we are doing now is a direct copy from GO, it maybe not necessary to add that comment? Please let me know your thought based on these info. I can add that if you prefer.

Reply:
Added.